modules/mobile-agents/android/client/src/org/wso2/mdm/agent/services/MessageProcessor.java

```diff
@@ -0,0 +1,134 @@
+import org.wso2.mdm.agent.utils.CommonUtils;
+
+import android.content.Context;
+import android.util.Log;
+
+/**
+ * Used to coordinate the retrieval and processing of messages from the server.
+ */
+public class MessageProcessor implements APIResultCallBack {
+
+	private String TAG = MessageProcessor.class.getSimpleName();
+	private Context context;
+	private String deviceId;
+	private static final String DEVICE_ID_PREFERENCE_KEY = "deviceId";
+	private static final String ENCODE_METHOD_UTF_8 = "utf-8";
+
+	/**
+	 * Local notification message handler.
+	 * @param context Context of the application.
+	 */
+	public MessageProcessor(Context context) {
+		this.context = context;
+
+		deviceId = Preference.getString(context, DEVICE_ID_PREFERENCE_KEY);
+		if (deviceId == null) {
+			DeviceInfo deviceInfo = new DeviceInfo(context.getApplicationContext());
+			deviceId = deviceInfo.getMACAddress();
+			Preference.putString(context, DEVICE_ID_PREFERENCE_KEY, deviceId);
+		}
+	}
+
+	/**
+	 * @param response Response received from the server that needs to be processed
+	 *                 and applied to the device.
+	 */
+	public void performOperation(String response) {
+		try {
+			JSONArray operations = new JSONArray(response);
+			for (int i = 0; i < operations.length(); i++) {
+				String featureCode = operations.getJSONObject(i).getString(Constants.CODE);
+				String properties = operations.getJSONObject(i).getString(Constants.PROPERTIES);
+				Operation operation = new Operation(context.getApplicationContext());
+				operation.doTask(featureCode, properties);
+			}
+		} catch (JSONException e) {
+			Log.e(TAG, "JSON Exception in response String." + e.toString());
+		}
+
+	}
+
+	/**
+	 * Call the message retrieval end point of the server to get messages
+	 * pending.
+	 */
+	public void getMessages() {
+		String ipSaved =
+				Preference.getString(context.getApplicationContext(),
+				                     context.getResources().getString(R.string.shared_pref_ip));
+		ServerConfig utils = new ServerConfig();
+		utils.setServerIP(ipSaved);
+		String deviceIdentifier = null;
+		
+		try {
+			deviceIdentifier = URLEncoder.encode(deviceId, ENCODE_METHOD_UTF_8);
+		} catch (UnsupportedEncodingException e) {
+			Log.e(TAG, "Unsupport encoding." + e.toString());
+		}
+		
+		CommonUtils.callSecuredAPI(context, utils.getAPIServerURL() +
+		                                    Constants.NOTIFICATION_ENDPOINT + File.separator +
+		                                    deviceIdentifier,
+		                           HTTP_METHODS.GET, new JSONObject(), MessageProcessor.this,
+		                           Constants.NOTIFICATION_REQUEST_CODE
+		);
+	}
+
+	@Override
+	public void onReceiveAPIResult(Map<String, String> result, int requestCode) {
+		String responseStatus;
+		String response;
+		if (requestCode == Constants.NOTIFICATION_REQUEST_CODE) {
+			if (result != null) {
+				responseStatus = result.get(Constants.STATUS_KEY);
+				if (responseStatus != null &&
```

Comment:
Compare the value assigned to variable "responseStatus" against the known constant Constants.REQUEST_SUCCESSFUL to avoid the preceding null check which is redundant.